---
title: "GSC RNAseq"
author: "Gurkan Bebek & Andrew Burr"
date: "December 4, 2015"
output: html_document
---

This is RNAseq analysis of the comparison between GSC vs differentiated NPC (10 vs. 10)  


```{r, cache=TRUE,echo=FALSE,eval=TRUE,warning=FALSE}




#loading required packages
require(knitr)
require(limma)
require(edgeR)
require(biomaRt)
require(ggplot2)
require(NMF)
require(DESeq2)
require(statmod)


library(fgsea)
library(data.table)
library(ggplot2)
require("reactome.db")



#function to annotate ensembl gene IDs to HGNC gene symbols
EnsembltoHGNC <- function(countsData){
  
  
  require(biomaRt)
  
  ensembl = useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "www.ensembl.org") # v82
  listMarts(mart=ensembl)
  # list all the Datasets in Ensembl biomart I connected
  listDatasets(ensembl) # 69 of them
  
  # I can also save Datasets available to me to a dataframe
  myDatasets<-listDatasets(ensembl)
  #View(myDatasets)
  # Ensembl currently contains >50 datasets~species
  
  # create a handle to connect Homo sapiens gene data
  ###n/a### ensembl = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
  ensembl = useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                    host = "www.ensembl.org", dataset="hsapiens_gene_ensembl")
  
  # find gene symbols for affy ids from chip HG-U133+v2
  #myAttributes<-listAttributes(ensembl)
  
  # or you can just use this to get all rows that have U133
  #myAttributes[grep("Ensembl",myAttributes$description),]
  
  
  # To build a query for biomart, you need 3 things:
  # provide attriutes (col names), filters (to filter rows), and values to query.
  annotation = getBM(attributes=c("ensembl_gene_id" , "hgnc_symbol" ), 
                     filters="ensembl_gene_id",  values=row.names(countsData),
                     mart = ensembl)
  
  countsData$ensembl_gene_id <- rownames(countsData)
  countsData <- merge(countsData, annotation, by = "ensembl_gene_id", all.x = T)
  #countsData <- countsData[countsData$hgnc_symbol != "",]
  #rownames(countsData) <- countsData$hgnc_symbol
  #countsData <- countsData[-c(1,22)]
  return(countsData)
  
}


#Function to import star counts from separate output folders
#requires the folder containing all the star counts and the column for use. Almost always column 2
importStarCounts <- function(folder, column){
  
  temp <- list.files(folder, full.names = TRUE)
  counts <- lapply(temp, read.delim, header=FALSE, stringsAsFactors = FALSE)
  file_names <- list.files(folder)
  
  counts <- lapply(counts, `[`, c(1,column))
  #turn each express output into a dataframe and merge them
  final_frame <- as.data.frame(counts[1])
  colnames(final_frame) <- c("ensembl_id", file_names[1])
  for(n in 2:length(counts)){
    temp_frame <- as.data.frame(counts[n])
    colnames(temp_frame) <- c("ensembl_id", file_names[n])
    final_frame <- merge(final_frame, temp_frame, by = "ensembl_id", all =TRUE)
    
  }
  rownames(final_frame) <- final_frame$ensembl_id
  final_frame <- subset(final_frame, select = -ensembl_id)
  #final_frame[is.na(finalFrame)] <- 0
  
  #final_frame <- final_frame[!final_frame$gene=="N_unmapped",]
  #final_frame <- final_frame[!final_frame$gene=="N_multimapping",]
  #final_frame <- final_frame[!final_frame$gene=="N_noFeature",]
  #final_frame <- final_frame[!final_frame$gene=="N_ambiguous",]
  
  
  return(final_frame)
}


#and this if from bowtie
importFeatureCounts <- function(countsFile){
  dataset <- read.delim(paste("CountsandExpression/" , countsFile, sep = ""), stringsAsFactors=FALSE)
  colnames(dataset) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++","387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", "1123++","528++","1123-","528-")
  dataset <- dataset[, c(1,2,3,4,5,6,7,8,17,18,9,10,11,12,13,14,15,16,19,20)]
  return(dataset)
}

conversion <- read.table("../../conversion_sets/ensembl_conversion.txt", header = T)


```




```{r GSC_vs_nonGSC}

#load in the star counts
read_counts <- importStarCounts("../star_v5/", column = 4)
colnames(read_counts) <- gsub(pattern = "_dna_ReadsPerGene.out.tab", 
                              "",
                              colnames(read_counts))
#write.table(test, "../data/read_counts.txt", quote = FALSE, row.names = F)


#change column names to the samples names form the data
colnames(read_counts) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++","387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", "1123++", "528++", "1123-", "528-", "hNP1_old","hNP1_new","17231","16157")

write.table(read_counts,
            "../data/all_read_counts.txt",
            sep = "\t",
            col.names = T)


#remove reads that did not get counted due to multi mapping and unalignment
read_counts <- read_counts[!rownames(read_counts)=="N_unmapped",]
read_counts <- read_counts[!rownames(read_counts)=="N_multimapping",]
read_counts <- read_counts[!rownames(read_counts)=="N_noFeature",]
read_counts <- read_counts[!rownames(read_counts)=="N_ambiguous",]


#select only the relevant sample sets
read_counts <- subset(read_counts,
                      select = -c(`1123++`, `528++`, `1123-`,`528-`,  hNP1_old, hNP1_new, `17231`, `16157` ))
#organize to that the non-GSCs are the first and the control samples. GSCs are the treatments
write.table(read_counts, 
            "../data/GSC_and_nonGSC_read_counts.txt",
            sep = "\t",
            col.names = T)

read_counts <- read_counts[,c(9:16, 1:8)]

#factor of the cell types
celltype <- factor(c(rep("D",8),   #D stands for differentiated cell or non-GSCs
                     rep("S",8)),   #S stands for stem cell of GSC
                   levels = c("D","S"))  

cellline <- factor(c(1:8, 1:8))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above
design <- model.matrix(~ cellline + celltype)   

#converting to DGEList
dge_counts <- DGEList(counts=read_counts[rowSums(cpm(read_counts) >= 1) >= 8,])
#dge_counts <- DGEList(counts=read_counts[rowSums(read_counts >= 10) >= 8,])

#normalization
dge_counts <- calcNormFactors(dge_counts,
                              method = c("TMM"))

#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(dge_counts, 
                   design, plot=TRUE)
#MDS plot of the data
plotMDS(dge_counts, 
        cex = 1,
        main = "MDS of Stem vs Diff",
        labels = colnames(read_counts),
        top = 500, 
        dim.plot = c(1,2),
        gene.selection = 'common')


#linear fit and eBayes funtions for differential expression
fit <-  eBayes(lmFit(voom_counts,
                     design))

#voom fold changes and p values for all genes
voom_all <- topTable(fit,
                     coef = "celltypeS", 
                     sort.by="p",
                     number =nrow(voom_counts))
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit,coef = "celltypeS",
                     sort.by="p",
                     number =nrow(voom_counts),
                     lfc = 0,
                     p.value = 0.05)


#name conversion to HGNC symbols
voom_all_HGNC <- EnsembltoHGNC(voom_all)
voom_top_HGNC <- EnsembltoHGNC(voom_top)


#saving the  top tables for use in other scripts
#commented out for use only when needed

write.table(x = voom_all_HGNC, 
            "../data/GSC_vs_nonGSC_all_differentially_expressed_genes.csv",
            quote = FALSE, 
            sep = ",", 
            row.names = T)
write.table(x = voom_top_HGNC,
            "../data/GSC_vs_nonGSC_top_pval05_differentially_expressed_genes.csv", 
            quote = FALSE, 
            sep = ",",
            row.names = T)
#write.table(x = as.data.frame(voom_counts$E), "../data/star_v5_normalized_counts.txt", quote = FALSE,  col.names = TRUE, sep = ",")





```


```{r fgsea}

conversion <- read.table("../../conversion_sets/ensembl_to_entrez.txt", stringsAsFactors = F, sep = "\t", header = T)
stats <- as.data.frame(voom_all_HGNC$logFC)
stats$hgnc_symbol <- voom_all_HGNC$hgnc_symbol
#stats <- stats[stats$symbol != "",]
stats <- merge(stats,conversion, by = 'hgnc_symbol')
ranks <- stats$`voom_all_HGNC$logFC`
stats$entrez_gene_id <- as.character(stats$entrez_gene_id)
names(ranks) <- stats$entrez_gene_id
ranks <- sort(ranks)

pathways <- reactomePathways(stats$entrez_gene_id)
test <- fgsea(pathways,exampleRanks)

data("examplePathways")
data("exampleRanks")
fgseaRes <- fgsea(pathways = pathways,
                  stats = ranks, 
                  nperm = 10000,
                  maxSize = 500)


pdf("../GSEA/GSC_vs_nonGSC_gsea.pdf", 14,8)
topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
print(
  
  plotGseaTable(pathways[topPathways], 
                ranks, 
                fgseaRes, 
                gseaParam = 0.5))
dev.off()




```




```{r GSC_vs_HNPC}

#load in the star counts
read_counts <- importStarCounts("../star_v5/", column = 4)

#remove reads that did not get counted due to multi mapping and unalignment
read_counts <- read_counts[!rownames(read_counts)=="N_unmapped",]
read_counts <- read_counts[!rownames(read_counts)=="N_multimapping",]
read_counts <- read_counts[!rownames(read_counts)=="N_noFeature",]
read_counts <- read_counts[!rownames(read_counts)=="N_ambiguous",]


#change column names to the samples names form the data
colnames(read_counts) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++","387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", "1123++", "528++", "1123-", "528-", "hNP1_old","hNP1_new","17231","16157")

#select only the relevant sample sets
#read_counts <- subset(read_counts, select = -c(`1123++`, `528++`, `1123-`,`528-`, `17231`, `16157` ))
read_counts <- subset(read_counts, 
                      select = -c(`387-`,`456-`, `3359-`,`3691-`,`3832-`,`4121-`,`4302-`,`H2S-`, `1123++`, `528++`, `1123-`, `528-`,`17231`,`16157`))


#organize to that the non-GSCs are the first and the control samples. GSCs are the treatments
read_counts <- read_counts[,c(9,10,1:8) ]
#factor of the cell types
celltype <- factor(c(rep("NPC",2),
                     rep("GSC",8)), 
                   levels = c("NPC","GSC"))
#cellline <- factor(c(1:8, 1:8, 9,9))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above
design <- model.matrix(~  celltype)   

#Voom recommends reads with low counts be removed
read_counts <- read_counts[rowSums(cpm(read_counts) >= 1) >= 2,]
#converting to DGEList
dge_counts <- DGEList(counts=read_counts)

#normalization
dge_counts <- calcNormFactors(dge_counts, 
                              method = c("TMM"))

#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(dge_counts, 
                   design, 
                   plot=TRUE)
#MDS plot of the data
plotMDS(dge_counts,
        cex = 1,
        main = "MDS of Stem vs Diff",
        labels = colnames(read_counts),
        top = 10000, 
        dim.plot = c(1,2), 
        gene.selection = 'common')


#linear fit and eBayes funtions for differential expression
fit <-  eBayes(lmFit(voom_counts,design))

#voom fold changes and p values for all genes
voom_all_genes <- topTable(fit, 
                           coef = "celltypeGSC",
                           sort.by="p",
                           number =nrow(voom_counts))
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit,coef = "celltypeGSC",
                     sort.by="p",
                     number =nrow(voom_counts),
                     lfc = 0,
                     p.value = 0.05)


#name conversion to HGNC symbols
voom_all_genes_HGNC <- EnsembltoHGNC(voom_all_genes)
voom_top_HGNC <- EnsembltoHGNC(voom_top)


#saving the  top tables for use in other scripts
#commented out for use only when needed



write.table(x = voom_all_HGNC, "../data/GSC_vs_HNPC_all_differentially_expressed_genes.csv",
            quote = FALSE,
            sep = ",",
            row.names = F)
write.table(x = voom_top_HGNC, 
            "../data/GSC_vs_HNPC_top_pval05_differentially_expressed_genes.csv",
            quote = FALSE, 
            sep = ",",
            row.names = F)
#write.table(x = as.data.frame(voom_counts$E), "../data/star_v5_normalized_counts.txt", quote = FALSE,  col.names = TRUE, sep = ",")



conversion <- read.table("../../conversion_sets/ensembl_to_entrez.txt", stringsAsFactors = F, sep = "\t", header = T)
stats <- as.data.frame(voom_all_HGNC$logFC)
stats$hgnc_symbol <- voom_all_HGNC$hgnc_symbol
#stats <- stats[stats$symbol != "",]
stats <- merge(stats,
               conversion,
               by = 'hgnc_symbol')
ranks <- stats$`voom_all_HGNC$logFC`
stats$entrez_gene_id <- as.character(stats$entrez_gene_id)
names(ranks) <- stats$entrez_gene_id
ranks <- sort(ranks)

pathways <- reactomePathways(stats$entrez_gene_id)
test <- fgsea(pathways,exampleRanks)

data("examplePathways")
data("exampleRanks")
fgseaRes <- fgsea(pathways = pathways,stats = ranks, nperm = 10000, maxSize = 500)


pdf("../GSEA/GSC_vs_HNPC_gsea.pdf", 14,8)
topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
print(
  
  plotGseaTable(pathways[topPathways], ranks, fgseaRes, 
                gseaParam = 0.5)
)
dev.off()





```



```{r RSEM}
#read in the RSEM counts
read_counts <- read.table("../RSEM_out/rsem_isoforms_matrix.txt")


#change column names to the samples names form the data
colnames(read_counts) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++","387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", "1123++", "528++", "1123-", "528-", "hNP1_old","hNP1_new","17231","16157")
read_counts <- subset(read_counts,
                      select = -c(`1123++`, `528++`, `1123-`,`528-`,  hNP1_old, hNP1_new, `17231`, `16157` ))
#organize to that the non-GSCs are the first and the control samples. GSCs are the treatments
read_counts <- read_counts[,c(9:16, 1:8)]
#factor of the cell types
celltype <- factor(c(rep("D",8),   #D stands for differentiated cell or non-GSCs
                     rep("S",8)),   #S stands for stem cell of GSC
                   levels = c("D","S"))

cellline <- factor(c(1:8, 1:8))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above
design <- model.matrix(~ cellline + celltype)   


#Voom recommends reads with low counts be removed

dge_counts <- DGEList(counts=read_counts[rowSums(cpm(read_counts) >= 1) >= 8,])

#normalization
dge_counts <- calcNormFactors(dge_counts, 
                              method = c("TMM"))

#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(dge_counts, design, plot=TRUE)
#MDS plot of the data
plotMDS(dge_counts,
        cex = 1, 
        main = "MDS of Stem vs Diff", 
        labels = colnames(read_counts), 
        top = 500, 
        dim.plot = c(1,2),
        gene.selection = 'common')


#linear fit and eBayes funtions for differential expression
fit <-  eBayes(lmFit(voom_counts,design))

#voom fold changes and p values for all genes
voom_all <- topTable(fit, 
                     coef = "celltypeS",
                     sort.by="p", 
                     number =nrow(voom_counts))
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit,
                     coef = "celltypeS", 
                     sort.by="p",
                     number =nrow(voom_counts),
                     lfc = 0,
                     p.value = 0.05)


#name conversion to HGNC symbols
voom_top$transcript_id <- rownames(voom_top)
voom_top_HGNC <- merge(voom_top, 
                       conversion[,c(2,3)], 
                       by = "transcript_id")

voom_all$transcript_id <- rownames(voom_all)
voom_all_HGNC <- merge(voom_all,
                       conversion[,c(2,3)], 
                       by = "transcript_id")


write.table(x = voom_all_HGNC,
            "../data/GSC_vs_nonGSC_all_differentially_expressed_transcripts.csv",
            quote = FALSE, 
            sep = ",",
            row.names = F)
write.table(x = voom_top_HGNC, 
            "../data/GSC_vs_nonGSC_top_pval05_differentially_expressed_transcripts.csv",
            quote = FALSE, 
            sep = ",",
            row.names = F)


```








```{r SRP040525}

#load in the star counts
read_counts <- importStarCounts("../SRP040525_counts/", column = 4)
colnames(read_counts) <- gsub(pattern = "ReadsPerGene.out.tab", "", colnames(read_counts))
#write.table(test, "../data/read_counts.txt", quote = FALSE, row.names = F)


#remove reads that did not get counted due to multi mapping and unalignment
read_counts <- read_counts[!rownames(read_counts)=="N_unmapped",]
read_counts <- read_counts[!rownames(read_counts)=="N_multimapping",]
read_counts <- read_counts[!rownames(read_counts)=="N_noFeature",]
read_counts <- read_counts[!rownames(read_counts)=="N_ambiguous",]
  

#select only the relevant sample sets
read_counts <- read_counts[,c(2,5,8,11)]

#factor of the cell types
kd <- factor(c(rep("WT",2),   #D stands for differentiated cell or non-GSCs
                     rep("KD",2)    #S stands for stem cell of GSC
                     #rep("C",2),
                    # rep("D",2)
                     ), levels = c("WT","KD")
                   )
celltype <- factor(c("S","D", "S","D"), levels = c("D", "S"))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above

#cellline <- factor(c(1:3,1:3), levels = 1:3)  #Factor for the samples used. 8 paired samples

#cell <- factor(paste(celltype, cellline, sep = "."))
#design <- model.matrix(~0+cell)
design <- model.matrix(~  kd + celltype)


#converting to DGEList
dge_counts <- DGEList(counts=read_counts[rowSums(cpm(read_counts) >= 1) >= 4,])

#normalization
dge_counts <- calcNormFactors(dge_counts, method = c("TMM"))

#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(dge_counts, design, plot=T)

#voom_counts <- cpm(dge_counts,log=T, prior.count = 3)

#MDS plot of the data
plotMDS(dge_counts, cex = 1, main = "MDS of Stem vs Diff", labels = colnames(read_counts), top = 500, dim.plot = c(1,2), gene.selection = 'common')

#fit <- lmFit(voom_counts,design)

#contrasts <- makeContrasts(((cellKD.2+cellWT.2)/2)-(cellKD.4+cellKD.4)/2  , levels = design)
#fit <- contrasts.fit(fit, contrasts)
#fit <- eBayes(fit, trend = T)

#linear fit and eBayes funtions for differential expression
fit <-  eBayes(lmFit(voom_counts,design))

#voom fold changes and p values for all genes
voom_all <- topTable(fit, coef = "celltypeS", sort.by="p", number =nrow(voom_counts))
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit,coef = "celltypeS", sort.by="p", number =nrow(voom_counts),lfc = 2,p.value = 0.05)


#name conversion to HGNC symbols
voom_all_HGNC <- EnsembltoHGNC(voom_all)
voom_top_HGNC <- EnsembltoHGNC(voom_top)


 #saving the  top tables for use in other scripts
#commented out for use only when needed

#write.table(x = voom_all_HGNC, "../data/StemVsDiff_all_DE_genes.txt", quote = FALSE, sep = ",", row.names = F)
#write.table(x = voom_top_HGNC, "../data/StemVsDiff_top_DE_genes.txt", quote = FALSE, sep = ",", row.names = F)
#write.table(x = as.data.frame(voom_counts$E), "../data/star_v5_normalized_counts.txt", quote = FALSE,  col.names = TRUE, sep = ",")





```


